tests/system/Database/Live/GroupTest.php

```diff
@@ -43,7 +43,7 @@ public function testHavingBy()
         $result = $this->db->table('job')
             ->select('name')
             ->groupBy('name')
-            ->having('SUM(id) > 2')
+            ->having('SUM(' . $this->db->protectIdentifiers('id') . ') > 2')
```

Comment:
I want to remove `$this->db->protectIdentifiers()`, because it is Query Builder.
But how?

Reply:
@kenjis 

Thanks, response.

You're right.

> But how?

Hmmm…

If you are using the OCI8 driver, what do you think about doing an ALTER TABLE to rename the id column to an uppercase column name?

This method does not essentially solve the QueryBuiler problem.
However, it may be a good solution in that it does not include this problem in the scope of this PR.

It looks just like this.

https://github.com/codeigniter4/CodeIgniter4/pull/2487/commits/729acdc7df399432bd2f5398c302b367cf019553#diff-c2d356f0d870b5a4a5bf49fad5e365c8ec6b3ee9b0f5a6bcb1f879801cd372d9L206-L215

Reply:
How about converting to uppercase when creating tables?

Reply:
No. Query Builder basically quotes identifier like `"id"`
when `$builder->where('id', 3);`

Reply:
@kenjis 

Thanks, response.

> No. Query Builder basically quotes identifiers like "id".

No. can’t escape identifier in parenthesis.

https://github.com/codeigniter4/CodeIgniter4/blob/4649204306276af6c06bfbd229fb1895fb5b0e4e/system/Database/BaseConnection.php#L970-L979

> How about converting to uppercase when creating tables?
> If you are using the OCI8 driver, what do you think about doing an ALTER TABLE to rename the id column to an uppercase column name?

I don't think I can solve either of these problems.
Oracle can handle lowercase letters as long as they are escaped.
However, if they are not escaped, they are treated as uppercase.
In other words, it will fail in the following cases.

https://github.com/codeigniter4/CodeIgniter4/pull/2487/files#diff-4d07390c9284f69a1dca0fd6dd95b0ba949997b35d1266828f0619bef65c7c22R58-R59

In the first having, the id is escaped and treated as lowercase.

The implementation I have in mind now is the following.

1. parse AST from SQL using a library like `PhpMyAdmin\SqlParser`.
2. escape identifiers from the AST.

In this case, we have the following concerns
I don't think it is desirable to run a heavy process many times for issuing a single query, what do you think?

Reply:
This is a issue in the Query Builder. There is no complete SQL statement to parse.
The current implementation of `protectIdentifiers()` may receive the string like a sub query `(SELECT MAX(advance_amount) FROM "orders" WHERE "id" > 2`, which does not have `)` at the end.

It seems to me that it is better to keep the specifications simple, maintain speed, and not do anything too complicated.

If we accept the comment as the specification of the Query Builder:
> If a parenthesis is found we know that we do not need to escape the data or add a prefix. 
> Added exception for single quotes as well, we don't want to alter literal strings.

The Query Builder never escapes items which contain `(`, `)`, and `'`.

In this case, it is recommended to write code like `->having('SUM("id") > 2')`.
No, `->having('SUM("id") >',  2)` is better.

Reply:
I think you are right.
However, MySQL uses the escape character "`".
Then my code will look like the following.

```php
$result = $this->db->table('job')
            ->select('name')
            ->groupBy('name')
            ->having('SUM(' . $this->db->protectIdentifiers('id') . ') >',  2);
```

In short, I wanted to erase `$this->db->protectIdentifiers()` the same way you did.
However, I can't think of a way to get rid of it.😭😭😭

Reply:
Okay. We can't solve this problem without making parser for parts of SQL (like `'SUM(id) > 2'`), 
and quotes identifiers automatically.

1. Use Quoted Identifiers
- When we create tables with *Forge* in a normal way, identifiers will be quoted.
- In this case, on Oracle, we always need to use quoted identifiers like `"name"`. 
- But MySQL uses `` ` `` to quote identifiers like `` `name` ``.

2. Use Unquoted Identifiers
- When we create tables with unquoted identifiers, Oracle considers them to be uppercase.
- Query Builder quotes identifiers in most cases.
- If we use a lowercase identifier like `name`, QB converts it to `"name"`, and there is no such identifier.

---

Oracle is a special case. I don't show code like `->having('SUM(' . $this->db->protectIdentifiers('id') . ') >',  2)`
only for Oracle.

How about using if statements to divide cases?

```php
    public function testHavingBy()
    {
        if ($this->db->DBDriver === 'OCI8') {
            $result = $this->db->table('job')
                ->select('name')
                ->groupBy('name')
                ->having('SUM("id") > 2')
                ->get()
                ->getResultArray();
        } else {
            $result = $this->db->table('job')
                ->select('name')
                ->groupBy('name')
                ->having('SUM(id) > 2')
                ->get()
                ->getResultArray();
        }

        $this->assertCount(2, $result);
    }
```

Reply:
@kenjis 

> How about using if statements to divide cases?

Very thanks, response.
I agree with this.
I'll modify it this way.

> only for Oracle.

I did not have enough knowledge about this.
I am sorry.

To be precise, it seems to be a specification of ANSI-SQL, not Oracle.
That is, it is ANSI-SQL compliant, and a similar implementation may be needed when implementing `Firebird`, which CodeIgniter will support.

Reply:
It seems you are correct. It comes from ANSI-SQL.

https://stackoverflow.com/questions/153944/is-sql-syntax-case-sensitive/19933159#19933159